judge/views/contests.py

```diff
@@ -1157,3 +1162,61 @@ def post(self, request, *args, **kwargs):
             return HttpResponseRedirect(self.get_success_url())
         else:
             return self.render_to_response(self.get_context_data(object=self.object))
+
+
+class ExportContestSubmissions(ContestMixin, SingleObjectMixin, View):
+    def get_object(self, queryset=None):
+        contest = super().get_object(queryset)
+        if not contest.is_editable_by(self.request.user):
+            raise PermissionDenied()
+        return contest
+
+    def get(self, request, *args, **kwargs):
+        contest = self.get_object()
+        output_dir = self.export_contest_submissions(contest)
+        response = HttpResponse(File(open(output_dir, 'rb')),
+                                content_type='application/zip')
+        os.remove(output_dir)
+        response['Content-Disposition'] = 'attachment; filename="%s-submissions.zip"' % (contest.key)
+        return response
+
+    def export_contest_submissions(self, contest):
+        output_dir = tempfile.mktemp(suffix='.zip')
+
+        users = contest.users.filter(virtual=ContestParticipation.LIVE).select_related('user__user')
+
+        submission_zip = zipfile.ZipFile(output_dir, 'w', zipfile.ZIP_DEFLATED)
+
+        for user in users:
+            username = user.user.user.username
+            user_dir = username
+            user_history_dir = os.path.join(user_dir, '$History')
+
+            problems = set()
+            submissions = user.submissions.order_by('-id') \
+                .select_related('problem__problem', 'submission__source', 'submission__language') \
+                .values_list('problem__problem__code', 'submission__source__source',
+                             'submission__language__extension', 'submission__id', 'language__file_only')
+
+            for problem, source, ext, sub_id, file_only in submissions:
+                if problem not in problems:  # Last submission
+                    problems.add(problem)
+                    if file_only:
+                        # Get the basename of the source as it is an URL
+                        filename = os.path.basename(source)
```

Comment:
The file_only url is structured as: `/submission_file/<contest_id>/<user_id>/<file_name>.<ext>`, so using os.path.basename will only get you the `<file_name>.<ext>` part (example: https://oj.vnoi.info/src/611645)